cmd/user/internal/application/eventhandler/when_user_was_authenticated_with_provider.go

```diff
@@ -0,0 +1,108 @@
+package eventhandler
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"log"
+
+	"github.com/vardius/go-api-boilerplate/cmd/user/internal/domain/user"
+	"github.com/vardius/go-api-boilerplate/cmd/user/internal/infrastructure/persistence"
+	"github.com/vardius/go-api-boilerplate/internal/domain"
+	"github.com/vardius/go-api-boilerplate/internal/eventbus"
+)
+
+// WhenUserWasAuthenticatedWithProvider handles event
+func WhenUserWasAuthenticatedWithProvider(db *sql.DB, repository persistence.UserRepository) eventbus.EventHandler {
+	fn := func(ctx context.Context, event domain.Event) {
+		// this goroutine runs independently to request's goroutine,
+		// there for recover middlewears will not recover from panic to prevent crash
+		defer recoverEventHandler()
+
+		log.Printf("[EventHandler] %s\n", event.Payload)
+
+		e := user.WasAuthenticatedWithProvider{}
+
+		err := json.Unmarshal(event.Payload, &e)
+		if err != nil {
+			log.Printf("[EventHandler] Error: %v\n", err)
+			return
+		}
+
+		tx, err := db.BeginTx(ctx, nil)
+		if err != nil {
+			log.Printf("[EventHandler] Error: %v\n", err)
+			return
+		}
+		defer tx.Rollback()
+
+		err = repository.Add(ctx, userWasAuthenticatedWithProviderModel{e})
+		if err != nil {
+			log.Printf("[EventHandler] Error: %v\n", err)
+			return
+		}
+
+		tx.Commit()
+	}
+
+	return fn
+}
+
+type userWasAuthenticatedWithProviderModel struct {
+	e user.WasAuthenticatedWithProvider
```

Comment:
why wrapping event with another type ?

Reply:
Saw it as a pattern used elsewhere and tried to follow coding style.

> https://github.com/vardius/go-api-boilerplate/blob/master/cmd/user/internal/application/eventhandler/when_user_was_registered_with_email.go

```
type userWasRegisteredWithEmailModel struct {
	e user.WasRegisteredWithEmail
}
```

Reply:
haha yeah